packages/examples/cvat/exchange-oracle/src/crons/_utils.py

```diff
@@ -0,0 +1,131 @@
+from collections.abc import Callable
+from contextlib import contextmanager, nullcontext
+from functools import wraps
+
+import httpx
+from sqlalchemy.orm import Session
+
+import src.services.webhook as oracle_db_service
+import src.services.webhook as webhook_service
+from src.core.types import OracleWebhookTypes
+from src.db import SessionLocal
+from src.db.utils import ForUpdateParams
+from src.models.webhook import Webhook
+from src.utils.webhooks import prepare_outgoing_webhook_body, prepare_signed_message
+
+
+def cron_job(logger_name: str) -> Callable[[Callable[..., None]], Callable[[], None]]:
+    """
+    Wrapper that supplies logger and optionally session to the cron job.
+
+    Example usage:
+        >>> @cron_job("app.cron.webhooks")
+        >>> def handle_webhook(logger: logging.Logger) -> None:
+        >>>     ...
+    Example usage with session:
+        >>> @cron_job("app.cron.webhooks")
+        >>> def handle_webhook(logger: logging.Logger, session: Session) -> None:
+        >>>     ...
+
+    Returns:
+        Cron job ready to be registered in scheduler.
+    """
+
+    def decorator(fn: Callable[..., None]):
+        logger = logging.getLogger(logger_name).getChild(fn.__name__)
+        cron_repr = repr(fn.__name__)
+
+        # validate signature
+        parameters = dict(inspect.signature(fn).parameters)
+        session_param = parameters.pop("session", None)
+        if session_param is not None:
+            assert (
+                session_param.annotation == Session
+            ), f"{cron_repr} session argument type muse be of type {Session.__qualname__}"
+        logger_param = parameters.pop("logger", None)
+        assert logger_param is not None, f"{cron_repr} must have logger argument"
+        assert (
+            logger_param.annotation == logging.Logger
+        ), f"{cron_repr} logger argument type muse be of type {logging.Logger}"
+        assert not parameters, (
+            f"{cron_repr} expected to have only have logger and session arguments,"
+            f" not {set(parameters.keys())}"
+        )
+
+        @wraps(fn)
+        def wrapper():
+            logger.debug(f"Cron {cron_repr} is starting")
+            try:
+                with SessionLocal.begin() if session_param else nullcontext() as session:
+                    return fn(logger, session) if session_param else fn(logger)
+            except Exception:
+                logger.exception(f"Exception while running {cron_repr} cron")
+            finally:
+                logger.debug(f"Cron {cron_repr} finished")
+
+        return wrapper
+
+    return decorator
+
+
+@contextmanager
+def handle_webhook(logger: logging.Logger, session: Session, webhook: Webhook):
+    savepoint = session.begin_nested()
+    logger.debug(
+        "Processing webhook "
+        f"{webhook.type}.{webhook.event_type}~{webhook.signature} "
+        f"in escrow_address={webhook.escrow_address} "
+        f"(attempt {webhook.attempts + 1})"
+    )
```

Comment:
I'm not entirely sure, what you mean when you say that logging can fail in this case. Could you elaborate on that?

In any case, I've rearranged these statements since it seemed logical to me after you pointed that out.

Reply:
The logging call, as any other call, can fail. In the implementation variant being commented, an exception in the logging call could lead to a session object not being closed properly.